auth/oidc.go

```diff
@@ -242,25 +246,34 @@ func (opm OIDCProviderMap) Stop() {
 	}
 }
 
-func (op *OIDCProvider) GetClient(buildCallbackURLFunc OIDCCallbackURLFunc) *OIDCClient {
-	// Initialize the client on first request. If the callback URL isn't defined for the provider,
-	// uses buildCallbackURLFunc to construct (based on current request)
-	op.clientOnce.Do(func() {
-		var err error
-		// If the redirect URL is not defined for the provider generate it from the
-		// handler request and set it on the provider
-		if op.CallbackURL == nil || *op.CallbackURL == "" {
-			callbackURL := buildCallbackURLFunc(op.Name, op.IsDefault)
-			if callbackURL != "" {
-				op.CallbackURL = &callbackURL
-			}
-		}
-		if err = op.initOIDCClient(); err != nil {
-			base.Errorf("Unable to initialize OIDC client: %v", err)
-		}
-	})
+// GetClient initializes the client on first successful request.
+func (op *OIDCProvider) GetClient(buildCallbackURLFunc OIDCCallbackURLFunc) (*OIDCClient, error) {
+	// If the callback URL isn't defined for the provider, uses buildCallbackURLFunc to construct (based on current request)
+
+	if op.clientInit.IsTrue() { // If client not initialized
+		return op.client, nil
+	}
+	// Acquire lock
+	op.clientInitLock.Lock()
+	defer op.clientInitLock.Unlock()
 
-	return op.client
+	if op.clientInit.IsTrue() { // Make sure past thread didn't init in lock
```

Comment:
This is a good example of a "why" comment :) but advise we keep comments on their own lines, it's easy to lose them when reading if they're tacked on to the end of code.

```suggestion
	// Check again to see if the previous lock holder initialized the client
	if op.clientInit.IsTrue() {
```